limits.go

```diff
@@ -15,143 +15,114 @@ import (
+		rcmgr.BaseLimitIncrease{StreamsInbound: 64, StreamsOutbound: 64, Streams: 128, Memory: 4 << 20},
+	)
+	config.AddServicePeerLimit(
+		identify.ServiceName,
+		rcmgr.BaseLimit{StreamsInbound: 16, StreamsOutbound: 16, Streams: 32, Memory: 1 << 20},
+		rcmgr.BaseLimitIncrease{},
+	)
+	for _, id := range [...]protocol.ID{identify.ID, identify.IDDelta, identify.IDPush} {
+		config.AddProtocolLimit(
+			id,
+			rcmgr.BaseLimit{StreamsInbound: 64, StreamsOutbound: 64, Streams: 128, Memory: 4 << 20},
+			rcmgr.BaseLimitIncrease{StreamsInbound: 64, StreamsOutbound: 64, Streams: 128, Memory: 4 << 20},
+		)
+		config.AddProtocolPeerLimit(
+			id,
+			rcmgr.BaseLimit{StreamsInbound: 16, StreamsOutbound: 16, Streams: 32, Memory: 32 * (256<<20 + 16<<10)},
+			rcmgr.BaseLimitIncrease{},
+		)
+	}
 
-	// ping
-	setServiceLimits(limiter, ping.ServiceName,
-		limiter.DefaultServiceLimits.
-			WithMemoryLimit(1, 4<<20, 64<<20). // max 64MB service memory
-			WithStreamLimit(128, 128, 128),    // max 128 streams - asymmetric
-		peerLimit(2, 3, 4))
-	setProtocolLimits(limiter, ping.ID,
-		limiter.DefaultProtocolLimits.WithMemoryLimit(1, 4<<20, 64<<20),
-		peerLimit(2, 3, 4))
+	//  ping
+	addServiceAndProtocolLimit(config,
+		ping.ServiceName, ping.ID,
+		rcmgr.BaseLimit{StreamsInbound: 64, StreamsOutbound: 64, Streams: 64, Memory: 4 << 20},
+		rcmgr.BaseLimitIncrease{StreamsInbound: 64, StreamsOutbound: 64, Streams: 64, Memory: 4 << 20},
+	)
+	addServicePeerAndProtocolPeerLimit(
+		config,
+		ping.ServiceName, ping.ID,
+		rcmgr.BaseLimit{StreamsInbound: 2, StreamsOutbound: 3, Streams: 4, Memory: 32 * (256<<20 + 16<<10)},
+		rcmgr.BaseLimitIncrease{},
+	)
 
 	// autonat
-	setServiceLimits(limiter, autonat.ServiceName,
-		limiter.DefaultServiceLimits.
-			WithMemoryLimit(1, 4<<20, 64<<20). // max 64MB service memory
-			WithStreamLimit(128, 128, 128),    // max 128 streams - asymmetric
-		peerLimit(2, 2, 2))
-	setProtocolLimits(limiter, autonat.AutoNATProto,
-		limiter.DefaultProtocolLimits.WithMemoryLimit(1, 4<<20, 64<<20),
-		peerLimit(2, 2, 2))
+	addServiceAndProtocolLimit(config,
+		autonat.ServiceName, autonat.AutoNATProto,
+		rcmgr.BaseLimit{StreamsInbound: 64, StreamsOutbound: 64, Streams: 64, Memory: 4 << 20},
+		rcmgr.BaseLimitIncrease{StreamsInbound: 4, StreamsOutbound: 4, Streams: 4, Memory: 2 << 20},
+	)
+	addServicePeerAndProtocolPeerLimit(
+		config,
+		autonat.ServiceName, autonat.AutoNATProto,
+		rcmgr.BaseLimit{StreamsInbound: 2, StreamsOutbound: 2, Streams: 2},
+		rcmgr.BaseLimitIncrease{},
+	)
 
 	// holepunch
-	setServiceLimits(limiter, holepunch.ServiceName,
-		limiter.DefaultServiceLimits.
-			WithMemoryLimit(1, 4<<20, 64<<20). // max 64MB service memory
-			WithStreamLimit(128, 128, 256),    // max 256 streams - symmetric
-		peerLimit(2, 2, 2))
-	setProtocolLimits(limiter, holepunch.Protocol,
-		limiter.DefaultProtocolLimits.WithMemoryLimit(1, 4<<20, 64<<20),
-		peerLimit(2, 2, 2))
+	addServiceAndProtocolLimit(config,
```

Comment:
I know this was what the old code did, but why do we need to set both the service and protocol limit? Shouldn’t the service limit cover the protocol? And/or shouldn’t the protocol limit cover the service?

Reply:
Yes, for services that use a single protocol. Identify uses multiple protocol IDs, so at least in theory, it could make sense to set separate limits.